### PR TITLE
Added support for .toPromise(PromiseConstructor)

### PR DESCRIPTION
**Pull request by Ginden**
I'm not sure about error message handling and tests.

Did you consider splitting `test/test.js` to many files?

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -1992,6 +1992,47 @@ addMethod('toCallback', function (cb) {
...
+ * @name Stream.toPromise(Promise)
+ * @param {Function} Promise - Promises/A+ compliant constructor
+ * @api public
+ *
+ * _([1, 2, 3, 4]).collect().toPromise(Promise).then(function (result) {
+ *     // parameter result will be [1,2,3,4]
+ * });
+ */
+
+addMethod('toPromise', function (Promise) {
```
Comment (vqvu) [inline lib/index.js]: Can you name the parameter `PromiseCtor` or something similar to highlight the fact that we do *not* use the global `Promise` object.

File: lib/index.js [hunk 2]
```diff
@@ -1992,6 +1992,47 @@ addMethod('toCallback', function (cb) {
 });
 
 /**
+ * Returns the result of a stream to a nodejs-style callback function.
```
Comment (vqvu) [inline lib/index.js]: Update docs.

You can just say the stream must emit at most one value or error and that the promise will resolve to that value when it is emitted, then link to `[toCallback](#toCallback)` for details.

**Comment by vqvu** [comment]
> error message

I would factor out the `toCallback` handler in such a way that you can customize the error message.

```javascript
function toCallbackHandler(transformName, cb) {
    var value;
    var hasValue = false; // In case an emitted value === null or === undefined.
    var hasValue = false; // In case an emitted value === null or === undefined.
 
    return function (err, x, push, next) {
        ...
        throws new Error(transformName + ' called on stream emitting multiple values');
        ...
    };
}

addMethod('toCallback', function (cb) {
    this.consume(toCallbackHandler('toCallback', cb)).resume();
});

addMethod('toPromise', function (PromiseCtor) {
    return new PromiseCtor(function(resolve, reject) {
        this.consume(toCallbackHandler('toPromise', function (err, x) {
            ...
        })).resume();
    });
});
```

> tests

The tests themselves look fine, but can you organize them into suites instead (e.g., like `exports.onDestroy`). This will make it easier to run the entire suite by itself.

```javascript
exports.toPromise = {
    ArrayStream: function ...
    'returns error for streams with multiple values': function ...
    ...
};
```

> Did you consider splitting `test/test.js` to many files?

Yep! I would love that; similarly with `lib/index.js`. It's just a lot of boring work, so no one has volunteered to do it yet.

**Comment by protometa** [comment]
What if PromiseCtor was optional and defaulted to global Promise? Best of both worlds?

**Comment by vqvu** [comment]
I had considered making `PromiseCtor`, and I didn't suggest it for two reasons.

1. We expose these methods at the top-level as [curried functions](http://highlandjs.org#currying), and currying does not work well with optional arguments, so we try to stay away from them. This is why the `sortBy` and `splitBy` variants exist as separate methods from `sort` and `split` rather than having the latter methods have an optional argument.

2. It doesn't solve the underlying problem, which is that it would be possible to write highland code that, but will behave differently depending on the Javascript environment. The way I see it, having `PromiseCtor` be optional would be no different than having two separate methods: `toPromise()` and `toPromiseWithCtor(PromiseCtor)`, and in such a scenario, I would still oppose to having `toPromise()`.

---

@Ginden, sorry I haven't responded to your latest changes. I don't get an email notification when you update the commits in your PR, so you have to make a comment. Are you planning on placing the tests in a separate `tests/toPromise.js` file? If so, please let me know when you are done.

**Comment by Ginden** [comment]
> . Are you planning on placing the tests in a separate tests/toPromise.js file? 

I do not plan it.

**Comment by vqvu** [comment]
Can you remove the file from the PR then?

**Comment by Ginden** [comment]
@vqvu Done.

**Comment by vqvu** [comment]
Thanks!

**Pull request merged**